core/src/replay_stage.rs

```diff
@@ -1171,12 +1174,247 @@ pub(crate) mod tests {
+        ];
+
+        let genesis_stake_accounts: Vec<Account> = vec![
+            stake_state::create_account(
+                &node_staking_keypairs[BOOTSTRAP_LEADER].pubkey(),
+                &node_voting_keypairs[BOOTSTRAP_LEADER].pubkey(),
+                &genesis_vote_accounts[BOOTSTRAP_LEADER],
+                &Rent::default(),
+                node_stakes[BOOTSTRAP_LEADER],
+            ),
+            stake_state::create_account(
+                &node_staking_keypairs[HONEST_NODE].pubkey(),
+                &node_voting_keypairs[HONEST_NODE].pubkey(),
+                &genesis_vote_accounts[HONEST_NODE],
+                &Rent::default(),
+                node_stakes[HONEST_NODE],
+            ),
+            stake_state::create_account(
+                &node_staking_keypairs[MALICIOUS_NODE].pubkey(),
+                &node_voting_keypairs[MALICIOUS_NODE].pubkey(),
+                &genesis_vote_accounts[MALICIOUS_NODE],
+                &Rent::default(),
+                node_stakes[MALICIOUS_NODE],
+            ),
+        ];
+
+        let mut genesis_config = create_genesis_config(10_000).genesis_config;
+        genesis_config.accounts.clear();
+
+        for i in BOOTSTRAP_LEADER..=MALICIOUS_NODE {
+            genesis_config.accounts.insert(
+                node_voting_keypairs[i].pubkey(),
+                genesis_vote_accounts[i].clone(),
+            );
+            genesis_config.accounts.insert(
+                node_staking_keypairs[i].pubkey(),
+                genesis_stake_accounts[i].clone(),
+            );
+        }
+
+        let mut bank_forks = BankForks::new(0, Bank::new(&genesis_config));
+
+        let mut fork_progress: HashMap<u64, ForkProgress> = HashMap::new();
+        fork_progress.entry(0).or_insert_with(|| {
+            ForkProgress::new(
+                bank_forks.banks[&0].slot(),
+                bank_forks.banks[&0].last_blockhash(),
+            )
+        });
+
+        for i in neutral_path.iter().skip(1) {
+            let bank = Bank::new_from_parent(
+                &bank_forks.banks[&(i - 1)],
+                &node_keypairs[BOOTSTRAP_LEADER].pubkey(),
+                *i,
+            );
+
+            bank.freeze();
+
+            fork_progress
+                .entry(*i)
+                .or_insert_with(|| ForkProgress::new(bank.slot(), bank.last_blockhash()));
+
+            bank_forks.insert(bank);
+
+            vote(
+                &bank_forks.banks[&i],
+                &node_voting_keypairs[BOOTSTRAP_LEADER].pubkey(),
+                *i - 1,
+            );
+
+            vote(
+                &bank_forks.banks[&i],
+                &node_voting_keypairs[HONEST_NODE].pubkey(),
+                *i - 1,
+            );
+
+            vote(
+                &bank_forks.banks[&i],
+                &node_voting_keypairs[MALICIOUS_NODE].pubkey(),
+                *i - 1,
+            );
+        }
+
+        let last_neutral_bank = &bank_forks.banks[&2].clone();
```

Comment:
this can be `bank_forks.banks.last().unwrap().clone()`

Reply:
@carllin HashMap doesn't have any ordering :)

Reply:
@carllin Better way to write this as `&bank_forks.banks[neutral_path.last().unwrap()].clone()`

Reply:
ah yeah, good point!